File: Android/app/src/main/java/uk/ewancroft/inkwell/ui/moderation/MutedBlockedViewModel.kt
package uk.ewancroft.inkwell.ui.moderation

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import uk.ewancroft.inkwell.data.repository.BlockedActorEntry
import uk.ewancroft.inkwell.data.repository.ModeratedActor
import uk.ewancroft.inkwell.data.repository.PdsRepository
import uk.ewancroft.inkwell.data.repository.fetchBlockedActors
import uk.ewancroft.inkwell.data.repository.fetchMutedActors
import uk.ewancroft.inkwell.data.repository.unmuteActor
import uk.ewancroft.inkwell.data.repository.deleteBlock
import javax.inject.Inject

data class MutedBlockedUiState(
    val isLoading: Boolean = true,
    val mutes: List<ModeratedActor> = emptyList(),
    val blocks: List<BlockedActorEntry> = emptyList(),
    val error: String? = null,
    /** DIDs/rkeys currently mid-removal, to disable their row and show a spinner. */
    val removingKeys: Set<String> = emptySet(),
)

/** Backs the "Muted & Blocked Accounts" screen (Settings → Account). */
@HiltViewModel
class MutedBlockedViewModel @Inject constructor(
    private val pdsRepository: PdsRepository,
) : ViewModel() {

    private val _uiState = MutableStateFlow(MutedBlockedUiState())
    val uiState: StateFlow<MutedBlockedUiState> = _uiState.asStateFlow()

    init {
        refresh()
    }

    fun refresh() {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(isLoading = true, error = null)
            try {
                val session = pdsRepository.getSession() ?: throw Exception("Not signed in")
                val mutes = pdsRepository.fetchMutedActors()
                val blocks = pdsRepository.fetchBlockedActors(session.did)
                _uiState.value = _uiState.value.copy(isLoading = false, mutes = mutes, blocks = blocks)
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(isLoading = false, error = e.message ?: "Failed to load.")
            }
        }
    }

    fun unmute(did: String) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(removingKeys = _uiState.value.removingKeys + did)
            try {
                pdsRepository.unmuteActor(did)
                _uiState.value = _uiState.value.copy(mutes = _uiState.value.mutes.filterNot { it.did == did })
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message ?: "Failed to unmute.")
            } finally {
                _uiState.value = _uiState.value.copy(removingKeys = _uiState.value.removingKeys - did)
            }
        }
    }

    fun unblock(entry: BlockedActorEntry) {
        viewModelScope.launch {
            _uiState.value = _uiState.value.copy(removingKeys = _uiState.value.removingKeys + entry.rkey)
            try {
                pdsRepository.deleteBlock(entry.rkey)
                _uiState.value = _uiState.value.copy(blocks = _uiState.value.blocks.filterNot { it.rkey == entry.rkey })
            } catch (e: Exception) {
                _uiState.value = _uiState.value.copy(error = e.message ?: "Failed to unblock.")
            } finally {
                _uiState.value = _uiState.value.copy(removingKeys = _uiState.value.removingKeys - entry.rkey)
            }
        }
    }
}

File: iOS/Inkwell/Features/Moderation/MutedBlockedViewModel.swift
//
//  MutedBlockedViewModel.swift
//  Inkwell
//

import Foundation
import Observation

@MainActor
@Observable
final class MutedBlockedViewModel {
    private let loginStateManager: LoginStateManager

    var isLoading = true
    var mutedActors: [ModeratedActor] = []
    var blockedActors: [BlockedActorEntry] = []
    var errorMessage: String?
    var removingKeys: Set<String> = []

    init(loginStateManager: LoginStateManager) {
        self.loginStateManager = loginStateManager
    }

    func load() async {
        isLoading = true
        errorMessage = nil
        do {
            let mutes = try await loginStateManager.fetchMutedActors()
            let blocks = try await loginStateManager.fetchBlockedActors()
            mutedActors = mutes
            blockedActors = blocks
        } catch {
            errorMessage = error.localizedDescription.isEmpty ? "Failed to load." : error.localizedDescription
        }
        isLoading = false
    }

    func unmute(did: String) async {
        guard !removingKeys.contains(did) else { return }
        removingKeys.insert(did)
        defer { removingKeys.remove(did) }

        do {
            try await loginStateManager.unmuteActor(did: did)
            mutedActors.removeAll { $0.did == did }
        } catch {
            errorMessage = error.localizedDescription.isEmpty ? "Failed to unmute." : error.localizedDescription
        }
    }

    func unblock(_ entry: BlockedActorEntry) async {
        guard !removingKeys.contains(entry.recordKey) else { return }
        removingKeys.insert(entry.recordKey)
        defer { removingKeys.remove(entry.recordKey) }

        do {
            try await loginStateManager.deleteBlock(recordKey: entry.recordKey)
            blockedActors.removeAll { $0.recordKey == entry.recordKey }
        } catch {
            errorMessage = error.localizedDescription.isEmpty ? "Failed to unblock." : error.localizedDescription
        }
    }
}
